### PR TITLE
Refactor doghouse / github-[pr-]check related code

### PR DESCRIPTION
**Pull request by haya14busa**
1. Implement GitHub Check API flow as CommentService / DiffService
2. Use the above {Comment,Diff}Service if reviewdog won't talk to
   the doghouse server (reviewdog GitHub App)
3. Implement github-pr-annotations reporter as {Comment,Diff}Service
4. It means doghouse related code (code under doghouse/ dir) should be used only when reviewdog
   needs to talk to the doghouse server

- [x] Updated Unreleased section in [CHANGELOG](https://github.com/reviewdog/reviewdog/blob/master/CHANGELOG.md) or it's not notable changes.

### CODE REVIEW

File: doghouse/server/doghouse.go [hunk 1]
```diff
@@ -34,18 +33,21 @@ const maxAnnotationsPerRequest = 50
...
 	return &Checker{
 		req:              req,
 		gh:               &checkerGitHubClient{Client: gh},
+		cli:              gh,
 		inDogHouseServer: inDogHouseServer,
 	}
 }
 
-func (ch *Checker) Check(ctx context.Context, makeRequest bool) (*doghouse.CheckResponse, error) {
+func (ch *Checker) Check(ctx context.Context) (*doghouse.CheckResponse, error) {
```
Comment (review-dog) [inline doghouse/server/doghouse.go:49]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method Checker.Check should have comment or be unexported

File: service/github/check.go [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+const maxAllowedSize = 65535
+
+// > The Checks API limits the number of annotations to a maximum of 50 per API
+// > request.
+// https://developer.github.com/v3/checks/runs/#output-object
+const maxAnnotationsPerRequest = 50
+
+var _ reviewdog.CommentService = (*Check)(nil)
+
+type Check struct {
```
Comment (review-dog) [inline service/github/check.go:34]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported type Check should have comment or be unexported

File: service/github/check.go [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+	Level    string
+
+	muComments   sync.Mutex
+	postComments []*reviewdog.Comment
+
+	muResult sync.Mutex
+	result   *CheckResult
+}
+
+type CheckResult struct {
```
Comment (review-dog) [inline service/github/check.go:50]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported type CheckResult should have comment or be unexported

File: service/github/check.go [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+	ReportURL  string
+	Conclusion string
+}
+
+func (ch *Check) Post(_ context.Context, c *reviewdog.Comment) error {
```
Comment (review-dog) [inline service/github/check.go:55]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method Check.Post should have comment or be unexported

File: service/github/check.go [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+	ch.muComments.Lock()
+	defer ch.muComments.Unlock()
+	ch.postComments = append(ch.postComments, c)
+	return nil
+}
+
+func (ch *Check) GetResult() *CheckResult {
```
Comment (review-dog) [inline service/github/check.go:62]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method Check.GetResult should have comment or be unexported

File: service/github/check.go [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+	ch.muResult.Lock()
+	defer ch.muResult.Unlock()
+	return ch.result
+}
+
+func (ch *Check) Flush(ctx context.Context) error {
```
Comment (review-dog) [inline service/github/check.go:68]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method Check.Flush should have comment or be unexported

File: service/github/check.go [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+		// Level takes precedence when configured (for backwards compatibility)
+		if len(annotations) == 0 {
+			return checkResult
+		}
+		switch strings.ToLower(ch.Level) {
+		case "info", "warning":
+			return "neutral"
+		}
+		return "failure"
+	} else {
```
Comment (review-dog) [inline service/github/check.go:227]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>if block ends with a return statement, so drop this else and outdent its block

File: service/github/diff.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	"os"
+	"os/exec"
+
+	"github.com/google/go-github/v60/github"
+	"github.com/reviewdog/reviewdog"
+)
+
+var _ reviewdog.DiffService = (*PullRequestDiffService)(nil)
+
+type PullRequestDiffService struct {
```
Comment (review-dog) [inline service/github/diff.go:17]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported type PullRequestDiffService should have comment or be unexported

File: doghouse/server/doghouse.go [hunk 9]
```diff
@@ -3,49 +3,37 @@ package server
...
```
Comment (review-dog) [inline doghouse/server/doghouse.go:35]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method Checker.Check should have comment or be unexported

File: service/github/check.go [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+const maxAllowedSize = 65535
+
+// > The Checks API limits the number of annotations to a maximum of 50 per API
+// > request.
+// https://developer.github.com/v3/checks/runs/#output-object
+const maxAnnotationsPerRequest = 50
+
+var _ reviewdog.CommentService = (*Check)(nil)
+
+type Check struct {
```
Comment (review-dog) [inline service/github/check.go:35]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported type Check should have comment or be unexported

File: service/github/check.go [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+	Level    string
+
+	muComments   sync.Mutex
+	postComments []*reviewdog.Comment
+
+	muResult sync.Mutex
+	result   *CheckResult
+}
+
+type CheckResult struct {
```
Comment (review-dog) [inline service/github/check.go:51]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported type CheckResult should have comment or be unexported

File: service/github/check.go [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+	ReportURL  string
+	Conclusion string
+}
+
+func (ch *Check) Post(_ context.Context, c *reviewdog.Comment) error {
```
Comment (review-dog) [inline service/github/check.go:56]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method Check.Post should have comment or be unexported

File: service/github/check.go [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+	ch.muComments.Lock()
+	defer ch.muComments.Unlock()
+	ch.postComments = append(ch.postComments, c)
+	return nil
+}
+
+func (ch *Check) GetResult() *CheckResult {
```
Comment (review-dog) [inline service/github/check.go:63]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method Check.GetResult should have comment or be unexported

File: service/github/check.go [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+	ch.muResult.Lock()
+	defer ch.muResult.Unlock()
+	return ch.result
+}
+
+func (ch *Check) Flush(ctx context.Context) error {
```
Comment (review-dog) [inline service/github/check.go:69]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method Check.Flush should have comment or be unexported

File: service/github/check.go [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+		// Level takes precedence when configured (for backwards compatibility)
+		if len(annotations) == 0 {
+			return checkResult
+		}
+		switch strings.ToLower(ch.Level) {
+		case "info", "warning":
+			return "neutral"
+		}
+		return "failure"
+	} else {
```
Comment (review-dog) [inline service/github/check.go:228]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>if block ends with a return statement, so drop this else and outdent its block

File: doghouse/server/doghouse.go [hunk 16]
```diff
@@ -3,49 +3,35 @@ package server
...
 	return &Checker{
 		req:              req,
-		gh:               &checkerGitHubClient{Client: gh},
+		cli:              gh,
 		inDogHouseServer: inDogHouseServer,
 	}
 }
 
-func (ch *Checker) Check(ctx context.Context, makeRequest bool) (*doghouse.CheckResponse, error) {
+func (ch *Checker) Check(ctx context.Context) (*doghouse.CheckResponse, error) {
```
Comment (review-dog) [inline doghouse/server/doghouse.go:33]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method Checker.Check should have comment or be unexported

File: service/github/check.go [hunk 17]
```diff
@@ -0,0 +1,347 @@
...
```
Comment (review-dog) [inline service/github/check.go:233]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>if block ends with a return statement, so drop this else and outdent its block

File: service/github/check.go [hunk 18]
```diff
@@ -0,0 +1,354 @@
...
+}
+
+func (ch *Check) Post(_ context.Context, c *reviewdog.Comment) error {
+	ch.muComments.Lock()
+	defer ch.muComments.Unlock()
+	ch.postComments = append(ch.postComments, c)
+	return nil
+}
+
+func (ch *Check) PostFiltered(_ context.Context, c *reviewdog.Comment) error {
```
Comment (review-dog) [inline service/github/check.go:63]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method Check.PostFiltered should have comment or be unexported

File: service/github/check.go [hunk 19]
```diff
@@ -0,0 +1,354 @@
...
+	ch.muComments.Lock()
+	defer ch.muComments.Unlock()
+	ch.postComments = append(ch.postComments, c)
+	return nil
+}
+
+func (ch *Check) GetResult() *CheckResult {
```
Comment (review-dog) [inline service/github/check.go:70]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method Check.GetResult should have comment or be unexported

File: service/github/check.go [hunk 20]
```diff
@@ -0,0 +1,354 @@
...
+	ch.muResult.Lock()
+	defer ch.muResult.Unlock()
+	return ch.result
+}
+
+func (ch *Check) Flush(ctx context.Context) error {
```
Comment (review-dog) [inline service/github/check.go:76]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method Check.Flush should have comment or be unexported

File: service/github/check.go [hunk 21]
```diff
@@ -0,0 +1,354 @@
...
+		// Level takes precedence when configured (for backwards compatibility)
+		if len(annotations) == 0 {
+			return checkResult
+		}
+		switch strings.ToLower(ch.Level) {
+		case "info", "warning":
+			return "neutral"
+		}
+		return "failure"
+	} else {
```
Comment (review-dog) [inline service/github/check.go:240]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>if block ends with a return statement, so drop this else and outdent its block

File: cmd/reviewdog/doghouse_test.go [hunk 22]
```diff
@@ -194,25 +104,22 @@ func TestDiagnosticResultSet_NonProject(t *testing.T) {
...
 		prNum = 14
 		sha   = "1414"
 	)
-
-	fakeCli := &fakeDoghouseServerCli{}
-	fakeCli.FakeCheck = func(ctx context.Context, req *doghouse.CheckRequest, makeRequest bool) (*doghouse.CheckResponse, error) {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/check", func(w http.ResponseWriter, r *http.Request) {
+		if r.Method != http.MethodPost {
+			t.Fatalf("unexpected method=%s", r.Method)
```
Comment (shogo82148) [inline cmd/reviewdog/doghouse_test.go:117]: We should not call `t.Fatalf` here.

```suggestion
			t.Errorf("unexpected method=%s", r.Method)
```

T.Fatalf calls T.FailNow internally, and [T.FailNow](https://pkg.go.dev/testing#T.FailNow) says:

> FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test.

This code will be called the other goroutine, so we should not use `t.Fatalf` here.

File: cmd/reviewdog/doghouse_test.go [hunk 23]
```diff
@@ -270,8 +177,19 @@ func TestPostResultSet_withReportURL(t *testing.T) {
...
+			ReportURL: "xxx",
+		}); err != nil {
+			t.Fatal(err)
+		}
+		w.WriteHeader(http.StatusOK)
+	})
+	ts := httptest.NewServer(mux)
+	defer ts.Close()
+
+	cli := client.New(&http.Client{})
```
Comment (shogo82148) [inline cmd/reviewdog/doghouse_test.go:191]: nit: we can use https://pkg.go.dev/net/http/httptest#Server.Client

```suggestion
	cli := client.New(ts.Client())
```

**Review by shogo82148 — APPROVED** [review]
LGTM

**Comment by haya14busa** [comment]
Thank you for your review!

**Pull request merged**